### PR TITLE
Add another client per API that can communicate with a custom named pipe

### PR DESCRIPTION
**Pull request by mauriciopoppe**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

There are two scenarios that we might see while deploying the CSI proxy:

- PD CSI driver imports the v1 CSI Proxy client, the node has the beta CSI Proxy server
- PD CSI driver imports the v1 CSI Proxy client, the node has the v1 CSI Proxy server

Instead of handling multiple imports in PD CSI driver we can just use the v1 client with additional startup code to identify which of these cases should attempt to solve.

This PR adds another way to initialize a client by sending a named pipe, the idea is that the user of the client can import the v1 client and still be able to call the v1beta named pipes for compatibility

The client is: 

```
func NewClientWithPipePath(pipePath string) (*Client, error)
```

The consumers of the client should send the `pipePath` they want to use, for example GCE PD CSI driver is using the following:

```golang
import (
	csiproxyclient "github.com/kubernetes-csi/csi-proxy/client"
	"github.com/kubernetes-csi/csi-proxy/client/apiversion"

	diskapi "github.com/kubernetes-csi/csi-proxy/client/api/disk/v1"
	diskclient "github.com/kubernetes-csi/csi-proxy/client/groups/disk/v1"
)

func getDiskClient() (*diskclient.Client, error) {
	var err error
	versionv1beta2 := apiversion.NewVersionOrPanic("v1beta2")
	diskAPIVersions := []apiversion.Version{diskClient.Version /* v1 */, versionv1beta2}

	// attempt to connect to one of the clients in order
	for _, apiVersion := range diskAPIVersions {
		pipePath := csiproxyclient.PipePath(diskclient.GroupName, apiVersion)
		klog.V(4).Infof("Attempting to connect to csi-proxy diskapi at path=%s", pipePath)
		diskClient, err := diskclient.NewClientWithPipePath(pipePath)
		if err == nil {
			klog.V(4).Infof("Connected to csi-proxy diskapi at path=%s", pipePath)
			return diskClient, nil
		}
		klog.V(4).Infof("Connection to csi-proxy diskapi at path=%s failed with error %v, might retry with a different version", pipePath, err)
	}
	return nil, fmt.Errorf("Couldn't connect to any csi-proxy diskapi server, last error %v", err)
}

func NewCSIProxyMounter() (*CSIProxyMounter, error) {
	diskClient, err := getDiskClient()
	if err != nil {
		return nil, err
	}

	fsClient, err := fsclient.NewClient()
	if err != nil {
		return nil, err
	}
	volumeClient, err := volumeclient.NewClient()
	if err != nil {
		return nil, err
	}
	return &CSIProxyMounter{
		FsClient:     fsClient,
		DiskClient:   diskClient,
		VolumeClient: volumeClient,
	}, nil
}
```

In my cluster I have CSI Proxy server v0.2, I've deployed GCE PD CSI driver with the v1 client (this change) and it could successfully avoid connecting to the v1 api because it doesn't exist, the logs:

```
I0407 05:51:58.024148    4288 safe-mounter_windows.go:63] Attempting to connect to csi-proxy diskapi at path=\\.\\pipe\\csi-proxy-disk-v1
I0407 05:51:58.024148    4288 safe-mounter_windows.go:69] Connection to csi-proxy diskapi at path=\\.\\pipe\\csi-proxy-disk-v1 failed with error=open \\.\\pipe\\csi-proxy-disk-v1: The system cannot find the file specified., might retry with a different version
I0407 05:51:58.025117    4288 safe-mounter_windows.go:63] Attempting to connect to csi-proxy diskapi at path=\\.\\pipe\\csi-proxy-disk-v1beta2
I0407 05:51:58.025117    4288 safe-mounter_windows.go:66] Connected to csi-proxy diskapi at path=\\.\\pipe\\csi-proxy-disk-v1beta2
I0407 05:51:58.025117    4288 safe-mounter_windows.go:81] Attempting to connect to csi-proxy fsapi at path=\\.\\pipe\\csi-proxy-filesystem-v1
I0407 05:51:58.025117    4288 safe-mounter_windows.go:87] Connection to csi-proxy fsapi at path=\\.\\pipe\\csi-proxy-filesystem-v1 failed with error=open \\.\\pipe\\csi-proxy-filesystem-v1: The system cannot find the file specified., might retry with a different version
I0407 05:51:58.025117    4288 safe-mounter_windows.go:81] Attempting to connect to csi-proxy fsapi at path=\\.\\pipe\\csi-proxy-filesystem-v1beta1
I0407 05:51:58.026099    4288 safe-mounter_windows.go:84] Connected to csi-proxy fsapi at path=\\.\\pipe\\csi-proxy-filesystem-v1beta1
I0407 05:51:58.026099    4288 safe-mounter_windows.go:99] Attempting to connect to csi-proxy volumeapi at path=\\.\\pipe\\csi-proxy-volume-v1
I0407 05:51:58.026099    4288 safe-mounter_windows.go:105] Connection to csi-proxy volumeapi at path=\\.\\pipe\\csi-proxy-volume-v1 failed with error=open \\.\\pipe\\csi-proxy-volume-v1: The system cannot find the file specified., might retry with a different version
I0407 05:51:58.026099    4288 safe-mounter_windows.go:99] Attempting to connect to csi-proxy volumeapi at path=\\.\\pipe\\csi-proxy-volume-v1beta1
I0407 05:51:58.026099    4288 safe-mounter_windows.go:102] Connected to csi-proxy volumeapi at path=\\.\\pipe\\csi-proxy-volume-v1beta1

```

Edit:

Even though I could connect to the v1beta API using the v1 client I couldn't send a message to it, the error I got was:

```
gce-pd-driver I0407 19:48:16.583101    8696 utils.go:60] /csi.v1.Node/NodeUnstageVolume called with request: volume_id:"projects/UNSPECIFIED/zones/us-central1-b/disks/pvc-1190e7da-72f1-480f-8c9d-95209b061660" staging_target_path:"\\var\\lib\\kubelet\\pl
 ugins\\kubernetes.io\\csi\\volumeDevices\\staging\\pvc-1190e7da-72f1-480f-8c9d-95209b061660"
 gce-pd-driver E0407 19:48:16.584111    8696 utils.go:63] /csi.v1.Node/NodeUnstageVolume returned with error: rpc error: code = Internal desc = NodeUnstageVolume failed: rpc error: code = Unimplemented desc = unknown service v1.Filesystem
```

In PD CSI we are falling back to using both APIs at the same time, in any case, the check with `winio.Dial` works :)

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/assign @jingxu97

### CODE REVIEW

File: client/utils.go [hunk 1]
```diff
@@ -17,3 +20,16 @@ const (
 func PipePath(apiGroupName string, apiVersion apiversion.Version) string {
 	return pipePrefix + csiProxyNamedPipePrefix + apiGroupName + "-" + apiVersion.String()
 }
+
+// FindFirstNamedPipe finds the first named pipe available at the Windows named pipe locations.
+// Only the first version from `apiVersions` that exists will be returned, this mechanism
+// allows picking the first version available in the case where there's a beta and v1 API
+func FindFirstNamedPipe(apiGroupName string, apiVersions []apiversion.Version) (string, error) {
```
Comment (jingxu97) [inline client/utils.go:27]: where this function is called?

Comment (mauriciopoppe) [reply client/utils.go:27]: in the code that's using the client (in our case PD CSI), please check out the description of the PR

Comment (mauriciopoppe) [reply client/utils.go]: I've removed this method and instead the scan is done in PD CSI

**Comment by ddebroy** [comment]
This is super cool!
/lgtm

**Comment by ddebroy** [comment]
Will approve shortly as soon as the `pull-kubernetes-csi-csi-proxy` reports

File: cmd/csi-proxy-api-gen/generators/client_generated_gen.go [hunk 2]
```diff
@@ -56,7 +58,19 @@ type Client struct {
...
+	pipePath := client.PipePath(GroupName, Version)
+	return NewClientWithPipePath(pipePath)
+}
+
+// NewClientWithPipePath returns a client to make calls to the named pipe located at "pipePath".
+// It's the caller's responsibility to Close the client when done.
+func NewClientWithPipePath(pipePath string) (*Client, error) {
+
+	// verify that the pipe exists
+	_, err := winio.DialPipe(pipePath, nil)
```
Comment (jingxu97) [inline cmd/csi-proxy-api-gen/generators/client_generated_gen.go:70]: wondering what is the difference between winio.DialPipe and the following grpc.Dial?

Comment (mauriciopoppe) [reply cmd/csi-proxy-api-gen/generators/client_generated_gen.go:70]: I read the first test in https://github.com/microsoft/go-winio/blob/master/pipe_test.go, `grpc.Dial` creates connections in the background unless `WithBlock()` is also sent as an option (which I also tried but got stuck in the v1 request)

**Pull request merged**